### PR TITLE
Add verification of signature (2.1.0)

### PR DESCRIPTION
**Pull request by perrettecl**
Signed-off-by: Clement Perrette <clementperrette@eaton.com>

### CODE REVIEW

File: tools/verify-fs.in [hunk 1]
```diff
@@ -94,9 +94,30 @@ for ROOTFS in *.squashfs; do
         log "Missing MD5 checksum file: '${ROOTFS}.md5'"
     fi
 
-    if [ -s "${ROOTFS}.sha256" ] ; then
+    if [ -s "${ROOTFS}.sha256" ] ; then
```
Comment (boricj) [inline tools/verify-fs.in]: Extraneous space at end of line.

File: tools/verify-fs.in [hunk 2]
```diff
@@ -94,9 +94,30 @@ for ROOTFS in *.squashfs; do
...
+                    SHA256_MANIFEST=$(grep ShaChecksum "${ROOTFS}-manifest.json" | cut -d '"' -f4)
+
+                if [ "${SHA256_COMPUTED}" = "${SHA256_MANIFEST}" ]; then
+                    if [ -s "${ROOTFS}-manifest.json.p7s" ] ; then
+                        SIGNED_BY=$(grep SignedBy "${ROOTFS}-manifest.json" | cut -d '"' -f4)
+
+                        if [ openssl cms -verify -binary -inform der -CAfile /usr/share/authorities/${SIGNED_BY}/rootCa1.pem -purpose any -out /dev/null -content "${ROOTFS}-manifest.json" -in "${ROOTFS}-manifest.json.p7s" ] ; then
+                            log "Image signature is valide for ${ROOTFS}"
+                        else
+                            log "Image signature is invalide for ${ROOTFS}"
```
Comment (boricj) [inline tools/verify-fs.in]: Typo: valide -> valid

File: tools/verify-fs.in [hunk 3]
```diff
@@ -94,9 +94,30 @@ for ROOTFS in *.squashfs; do
...
-            log "Good integrity of ${ROOTFS}, SHA256 checksum is valid"
+            if [ -s "${ROOTFS}-manifest.json" ] ; then
+                    SHA256_COMPUTED=$(cut -d ' ' -f1 < "${ROOTFS}.sha256")
+                    SHA256_MANIFEST=$(grep ShaChecksum "${ROOTFS}-manifest.json" | cut -d '"' -f4)
+
+                if [ "${SHA256_COMPUTED}" = "${SHA256_MANIFEST}" ]; then
+                    if [ -s "${ROOTFS}-manifest.json.p7s" ] ; then
+                        SIGNED_BY=$(grep SignedBy "${ROOTFS}-manifest.json" | cut -d '"' -f4)
+
+                        if [ openssl cms -verify -binary -inform der -CAfile /usr/share/authorities/${SIGNED_BY}/rootCa1.pem -purpose any -out /dev/null -content "${ROOTFS}-manifest.json" -in "${ROOTFS}-manifest.json.p7s" ] ; then
```
Comment (boricj) [inline tools/verify-fs.in]: `CAfile` parameter should be double-quoted.

File: tools/verify-fs.in [hunk 4]
```diff
@@ -94,9 +94,30 @@ for ROOTFS in *.squashfs; do
...
+                            log "Image signature is valide for ${ROOTFS}"
+                        else
+                            log "Image signature is invalide for ${ROOTFS}"
+                        fi
+                    else
+                        log "Manifest signature not found for ${ROOTFS}"
+                    fi
+                else
+                    log "Manifest sha256 do not match with the sha256 for ${ROOTFS}"
```
Comment (boricj) [inline tools/verify-fs.in]: Better message would be `Manifest sha256 hash does not match with the sha256 hash for ${ROOTFS}`

**Comment by aquette** [comment]
> Parsing JSON with `grep` and `cut` seems very hazardous... Try finding a more robust solution.

we've JSON.sh in place, at least

File: tools/verify-fs.in [hunk 5]
```diff
@@ -95,25 +95,36 @@ for ROOTFS in *.squashfs; do
...
+            fi
         else
             log "Bad integrity of ${ROOTFS}, SHA256 checksum is not valid"
         fi
     else
         log "Missing SHA256 checksum file: '${ROOTFS}.sha256'"
     fi
 
-    if [ -s "${ROOTFS}.cksum" ] ; then
-        if [ /usr/bin/cksum -c "${ROOTFS}.cksum" 2>/dev/null ] ; then
```
Comment (jimklimov) [inline tools/verify-fs.in]: Note: removing this block misses out on size-verification of cksum (we might also use the size embedded to manifest).

But notably our images have a version of this tool that does not support `cksum -c` so this block would need at least fixing to pass like done in other places (call `cksum FILENAME` and pick first two columns of output, compare to the two in saved cksum file).

File: tools/verify-fs.in [hunk 6]
```diff
@@ -95,25 +95,36 @@ for ROOTFS in *.squashfs; do
     fi
 
     if [ -s "${ROOTFS}.sha256" ] ; then
-        if [ /usr/bin/sha256sum -c "${ROOTFS}.sha256" 2>/dev/null ] ; then
-            log "Good integrity of ${ROOTFS}, SHA256 checksum is valid"
+        if /usr/bin/sha256sum -c "${ROOTFS}.sha256" 2>/dev/null; then
+            if [ -s "${ROOTFS}-manifest.json" ] ; then
+                    SHA256_COMPUTED=$(cut -d ' ' -f1 < "${ROOTFS}.sha256")
+                    SHA256_MANIFEST=$(/usr/share/fty/scripts/JSON.sh -l < "${ROOTFS}-manifest.json" | grep '\["application","ShaChecksum"\]' | cut -f2 | cut -d '"' -f2)
```
Comment (jimklimov) [inline tools/verify-fs.in]: Note: JSON processing here can be much streamlined and simplified (for consumer call syntax) by using the updated tool version from master's PR #408

**Pull request merged**